### PR TITLE
Remember the last selected node in topology

### PR DESCRIPTION
**Pull request by vikram-raj**
**Story:** https://issues.redhat.com/browse/ODC-6053

**Descriptions:**
In Topology remember the last opened overview panel by the user.

**GIF:**
Kapture 2021-11-24 at 21 01 49

### CODE REVIEW

**Comment by nemesis09** [comment]
tried locally. works as expected.
/lgtm

**Comment by serenamarie125** [comment]
This looks great - can we verify that if we remember a selected node in project-a, it won't cause issues when changing to project-b & going to topology?  You may have already tested this, but don't see it in the animated gif.  Thank you!

**Comment by nemesis09** [comment]
@serenamarie125 I tested this out. On changing the project, no node is selected by default and on selecting a node, the selection is remembered until we change the project again. The screen recording is attached.

https://user-images.githubusercontent.com/38663217/143387856-372b1768-8660-4f55-bd93-42dd657a3131.mp4

**Comment by vikram-raj** [comment]
Propagate from epic:
/label docs-approved
/label px-approved

File: frontend/packages/topology/src/components/page/TopologyView.tsx [hunk 1]
```diff
@@ -160,16 +161,26 @@ export const ConnectedTopologyView: React.FC<ComponentProps> = ({
...
+      // set empty selection when selecting the graph
+      const selEntity = isGraph(entity) ? undefined : entity;
+      setSelectedEntity(selEntity);
+      if (!selEntity) {
+        removeQueryArgument('selectId');
+      } else {
+        setQueryArgument('selectId', selEntity.getId());
+        sessionStorage.setItem(
+          LAST_TOPOLOGY_OVERVIEW_OPEN_STORAGE_KEY,
+          JSON.stringify({ [namespace]: selEntity.getId() }),
```
Comment (jerolimov) [inline frontend/packages/topology/src/components/page/TopologyView.tsx:178]: This removes the selected entry from another project which IMHO doesn't align with the AC "per project".

I would recommend to parse and spread the old values from sessionStorage

Comment (vikram-raj) [reply frontend/packages/topology/src/components/page/TopologyView.tsx:178]: PM requirement for this [story](https://issues.redhat.com/browse/ODC-6053) is to just remember the node selection for the last active namespace. cc: @serenamarie125

Comment (jerolimov) [reply frontend/packages/topology/src/components/page/TopologyView.tsx:178]: AC 1 sounds like we should save it for all namespaces. AC 2 is not added by PM. So let see if we get some feedback from Serena here. :)

@serenamarie125 @beaumorley I'm talking about this case:

* User selects a Deployment in project1
* User switch to project2 to check something, selects another Deployment
* User switch back to project1. Should we automatically select the first Deployment again?

The change will be straightforward, don't worry about that.

**Review by jerolimov — CHANGES_REQUESTED** [review]
@vikram-raj Please also add some unit and/or e2e tests for this new feature.

**Comment by serenamarie125** [comment]
@jerolimov imo, it is not necessary to remember the node selection from a "previous" project.

**Comment by vikram-raj** [comment]
@jerolimov  e2e tests for this feature has been added to PR https://github.com/openshift/console/pull/10413

File: frontend/packages/topology/src/components/page/TopologyView.tsx [hunk 2]
```diff
@@ -160,16 +161,26 @@ export const ConnectedTopologyView: React.FC<ComponentProps> = ({
...
+    (entity?: GraphElement) => {
+      if (!entity) {
+        sessionStorage.removeItem(LAST_TOPOLOGY_OVERVIEW_OPEN_STORAGE_KEY);
+      }
+      // set empty selection when selecting the graph
+      const selEntity = isGraph(entity) ? undefined : entity;
+      setSelectedEntity(selEntity);
+      if (!selEntity) {
+        removeQueryArgument('selectId');
+      } else {
```
Comment (jerolimov) [inline frontend/packages/topology/src/components/page/TopologyView.tsx:174]: I think your `sessionStorage.removeItem` should be part of the `if (!selEntity)` check or? At least the comment sounds like that to me.
```suggestion
      // set empty selection when selecting the graph
      const selEntity = isGraph(entity) ? undefined : entity;
      setSelectedEntity(selEntity);
      if (!selEntity) {
        sessionStorage.removeItem(LAST_TOPOLOGY_OVERVIEW_OPEN_STORAGE_KEY);
        removeQueryArgument('selectId');
      } else {
```

**Review by jerolimov — CHANGES_REQUESTED** [review]

File: frontend/packages/topology/src/components/page/TopologyView.tsx (hunk 2 above)
Comment (vikram-raj) [reply frontend/packages/topology/src/components/page/TopologyView.tsx]: Yes, my bad. I updated it. PTAL. Thanks

File: frontend/packages/topology/src/components/page/TopologyPage.tsx [hunk 3]
```diff
@@ -100,6 +101,15 @@ export const TopologyPage: React.FC<TopologyPageProps> = ({
...
     (queryParams.get('view') as TopologyViewType) || topologyViewState || defaultViewType;
 
+  React.useEffect(() => {
+    const lastOverviewOpen = JSON.parse(
+      sessionStorage.getItem(LAST_TOPOLOGY_OVERVIEW_OPEN_STORAGE_KEY) ?? '{}',
+    );
+    if (loaded && namespace in lastOverviewOpen) {
+      setQueryArgument('selectId', lastOverviewOpen[namespace]);
+    }
+  });
```
Comment (jerolimov) [inline frontend/packages/topology/src/components/page/TopologyPage.tsx:111]: Please add these deps here. With this two the hook was only called once. Without them it was called whenever I select a topology node.

```suggestion
  }, [loaded, namespace]);
```

File: frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx [hunk 4]
```diff
@@ -123,7 +123,12 @@ const QuickSearchList: React.FC<QuickSearchListProps> = ({
                       wrapModifier="truncate"
                       key={`${item.uid}-name`}
                     >
-                      <span className="ocs-quick-search-list__item-name">{item.name}</span>
+                      <span
+                        className="ocs-quick-search-list__item-name"
+                        data-test={`item-name-${item.name}-${itemType}`}
```
Comment (jerolimov) [inline frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx:128]: Not used `data-test` attribute? Is it needed for another PR or could we drop it?

**Review by jerolimov — CHANGES_REQUESTED** [review]
Small nits we already discussed in slack.

File: frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx (hunk 4 above)
Comment (vikram-raj) [reply frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx]: Not needed.

File: frontend/packages/topology/src/components/page/TopologyPage.tsx (hunk 3 above)
Comment (vikram-raj) [reply frontend/packages/topology/src/components/page/TopologyPage.tsx]: Updated it.

**Review by jerolimov — APPROVED** [review]
Tested and works fine. :+1: 

(Tested in admin and dev perspective with Chrome. I will add qe-approved label also after running it once on a cluster as well. A cluster is spinning up at the moment.)

/lgtm
/approve

**Comment by jerolimov** [comment]
Tested this on a cluster and works fine.

/label qe-approved

**Pull request merged**